File: static/CWE-78/c/vulnerable.c
/* CWE-78: Improper Neutralization of Special Elements used in an OS Command ('OS Command Injection')
   A shell command string is built from untrusted argv and executed.
   Vulnerable sink: system() */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char **argv)
{
    char command[256];

    if (argc < 2)
        return 1;

    /* Untrusted input concatenated into a shell command */
    snprintf(command, sizeof(command), "ping -c 1 %s", argv[1]);
    system(command);

    return 0;
}

File: static/CWE-78/javascript/vulnerable.js
// CWE-78: Improper Neutralization of Special Elements used in an OS Command ('OS Command Injection')
// Untrusted request input is interpolated into a shell command executed by the OS.
// Vulnerable sink: child_process.exec(`ping -c 1 ${req.query.host}`)

const express = require('express');
const { exec } = require('child_process');

const app = express();

app.get('/ping', (req, res) => {
  const host = req.query.host;
  exec(`ping -c 1 ${host}`, (err, stdout, stderr) => {
    if (err) return res.status(500).send(stderr);
    res.send(stdout);
  });
});

app.listen(3000);
